com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/DeleteServicesStep.java

```diff
@@ -1,124 +1,232 @@
+                                                                     binding.getApplicationGuid()));
+            }
+            getStepLogger().info(Messages.UNBINDING_SERVICE_FROM_APP, serviceName, application.getName());
+            client.unbindService(application.getName(), serviceName);
+        }
+    }
+
+    private void deleteServiceKeys(CloudControllerClient client, String serviceName) {
+        CloudService service = client.getService(serviceName);
+        if (service.isUserProvided()) {
+            return;
+        }
+        List<CloudServiceKey> serviceKeys = client.getServiceKeys(serviceName);
+        for (CloudServiceKey serviceKey : serviceKeys) {
+            getStepLogger().info(Messages.DELETING_SERVICE_KEY_FOR_SERVICE, serviceKey.getName(), serviceName);
+            client.deleteServiceKey(serviceName, serviceKey.getName());
+        }
+    }
+
+    private void deleteService(CloudControllerClient client, String serviceName) {
+        getStepLogger().info(Messages.DELETING_SERVICE, serviceName);
+        client.deleteService(serviceName);
+        getStepLogger().debug(Messages.SERVICE_DELETED, serviceName);
+    }
+
+    private void processException(DelegateExecution context, Exception e, CloudServiceInstance serviceInstance, String serviceName) {
+        if (e instanceof CloudOperationException) {
+            e = evaluateCloudOperationException(context, (CloudOperationException) e, serviceName, serviceInstance.getService()
+                                                                                                                  .getLabel());
+            if (e == null) {
+                return;
+            }
+        }
+        wrapAndThrowException(e, serviceInstance, serviceName);
+    }
+
+    private CloudOperationException evaluateCloudOperationException(DelegateExecution context, CloudOperationException e,
+                                                                    String serviceName, String label) {
+        if (e.getStatusCode() == HttpStatus.NOT_FOUND) {
+            getStepLogger().warn(MessageFormat.format(Messages.COULD_NOT_DELETE_SERVICE, serviceName), e,
+                                 ExceptionMessageTailMapper.map(configuration, CloudComponents.SERVICE_BROKERS, label));
+            return null;
+        }
+        if (e.getStatusCode() == HttpStatus.BAD_GATEWAY) {
+            StepsUtil.setServiceOffering(context, Constants.VAR_SERVICE_OFFERING, label);
+            return new CloudServiceBrokerException(e);
+        }
+        return new CloudControllerException(e);
+
+    }
+
+    private void wrapAndThrowException(Exception e, CloudServiceInstance serviceInstance, String serviceName) {
+        String msg = buildNewExceptionMessage(e, serviceInstance, serviceName);
+        throw new SLException(e, msg);
+    }
+
+    private String buildNewExceptionMessage(Exception e, CloudServiceInstance serviceInstance, String serviceName) {
```

Comment:
The old comment still stands.
The method name is not very descriptive. Reading it, does not point at what exception message exactly it builds, especially the 'New' keyword. I had to check in the implementation to understand that it builds an error message wen deleting a service.